src/Microsoft.Diagnostics.Monitoring.StartupHook/Exceptions/Pipeline/Steps/ExceptionIdSource.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Threading;
+
+namespace Microsoft.Diagnostics.Monitoring.StartupHook.Exceptions.Pipeline.Steps
+{
+    internal sealed class ExceptionIdSource
+    {
+        private readonly object _exceptionIdKey = new object();
+
+        private ulong _nextExceptionId = 1;
+
+        public ulong GetId(Exception? exception)
+        {
+            if (null == exception)
+                return 0;
+
+            if (TryGetExceptionId(exception, out ulong exceptionId))
+                return exceptionId;
+
+            lock (exception.Data)
+            {
+                if (TryGetExceptionId(exception, out exceptionId))
```

Comment:
I'm assuming I'm missing something here (and I see we had this logic before) - why do we call this method twice in a row, once without the lock and once with the lock?

Reply:
In case some other thread assigned the exception an identifier while it held the lock and this thread was waiting on the lock; not likely to happen, but can happen if an exception is created on one thread, and thrown from multiple threads at the same time.